app/org/sagebionetworks/bridge/cache/CacheProvider.java

```diff
@@ -19,62 +27,115 @@
 @Component
 public class CacheProvider {
 
+    private ObjectMapper bridgeObjectMapper;
     private JedisOps jedisOps;
 
     @Autowired
+    public void setBridgeObjectMapper(BridgeObjectMapper bridgeObjectMapper) {
+        this.bridgeObjectMapper = bridgeObjectMapper;
+    }
+
+    @Autowired
     public void setJedisOps(JedisOps jedisOps) {
         this.jedisOps = jedisOps;
     }
 
-    public void setUserSession(String key, UserSession session) {
+    public void setUserSession(final UserSession session) {
+
+        checkNotNull(session);
+        final User user = session.getUser();
+        checkNotNull(user, "Missing user in session.");
+        final String userId = user.getId();
+        checkNotNull(userId, "Missing user ID in session.");
+        final String sessionToken = session.getSessionToken();
+        checkNotNull(sessionToken, "Missing session token for session.");
+
+        final String userKey = RedisKey.USER_SESSION.getRedisKey(userId);
+        final String sessionKey = RedisKey.SESSION.getRedisKey(sessionToken);
         try {
-            String ser = BridgeObjectMapper.get().writeValueAsString(session);
-            String redisKey = RedisKey.SESSION.getRedisKey(key);
-            String result = jedisOps.setex(redisKey, BridgeConstants.BRIDGE_SESSION_EXPIRE_IN_SECONDS, ser);
-            if (!"OK".equals(result)) {
-                throw new BridgeServiceException("Session storage error");
+            final String ser = bridgeObjectMapper.writeValueAsString(session);
+            final List<Object> results = jedisOps.getTransaction()
+                    .setex(userKey, BridgeConstants.BRIDGE_SESSION_EXPIRE_IN_SECONDS, sessionToken)
+                    .setex(sessionKey, BridgeConstants.BRIDGE_SESSION_EXPIRE_IN_SECONDS, ser)
+                    .exec();
+            if (results == null) {
+                throw new BridgeServiceException("Session storage error.");
             }
-        } catch (Throwable e) {
-            promptToStartRedisIfLocalEnv(e);
+            for (Object result : results) {
+                if (!"OK".equals(result)) {
+                    throw new BridgeServiceException("Session storage error.");
+                }
+            }
+        } catch (JsonProcessingException e) {
+            promptToStartRedisIfLocal(e);
             throw new BridgeServiceException(e);
         }
     }
 
-    public UserSession getUserSession(String key) {
+    public UserSession getUserSession(final String sessionToken) {
+        final String sessionKey = RedisKey.SESSION.getRedisKey(sessionToken);
+        final String ser = jedisOps.get(sessionKey);
```

Comment:
Jedis exceptions will be wrapped into BridgeServiceException by downstream service/controller code.  Behind services and controllers we can just pass whatever exception it is.  Actually does it make sense to remove all the try-catch-rethrow in CacheProvider? What do you think, Alx?